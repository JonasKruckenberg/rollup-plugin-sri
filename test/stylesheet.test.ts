import { getHtml, emitAsset } from './util'
import { rollup } from 'rollup'
import { join } from 'path'
import { readFile } from 'fs/promises'
import sri from '../index'

process.chdir(join(__dirname, 'fixtures'));

describe('stylesheet', () => {
  test('basic', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
              <link rel="stylesheet" href="1.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('remote', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="https://raw.githubusercontent.com/JonasKruckenberg/rollup-plugin-sri/master/test/fixtures/1.css"></link>
          </body>
        </html>`),
        sri()
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('multiple hash algorithms', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="1.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri({ algorithms: ['sha256', 'sha384', 'sha512', 'whirlpool'] })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('uppercase hash algorithms', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="1.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri({ algorithms: ['SHA256', 'SHA384', 'SHA512'] })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('invalid hash algorithm', async () => {
    const onwarn = jest.fn()

    const bundle = await rollup({
      input: '1.js',
      onwarn,
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="1.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri({ algorithms: ['sha1'] })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
    expect(onwarn.mock.calls).toHaveLength(1)
    expect(onwarn.mock.calls[0][0]).toHaveProperty('message','Insecure hashing algorithm "sha1" will be rejected by browsers!')
  })

  test('crossorigin attribute', async () => {
    const bundle = await rollup({
      input: '1.js',
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="1.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri({ crossorigin: 'use-credentials' })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
  })

  test('invalid link', async () => {
    const onwarn = jest.fn()

    const bundle = await rollup({
      input: '1.js',
      onwarn,
      plugins: [
        emitAsset('index.html', `
        <!DOCTYPE html>
          <html lang="en">
          <body>
            <link rel="stylesheet" href="invalid.css"></link>
          </body>
        </html>`),
        emitAsset('1.css', await readFile('1.css', 'utf-8')),
        sri({ crossorigin: 'use-credentials' })
      ]
    })

    const source = await getHtml(bundle)
    expect(source).toMatchSnapshot()
    expect(onwarn.mock.calls).toHaveLength(1)
    expect(onwarn.mock.calls[0][0]).toHaveProperty('message','could not resolve resource "invalid.css"!')  
  })
})