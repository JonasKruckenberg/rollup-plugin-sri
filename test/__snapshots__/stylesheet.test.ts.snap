// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stylesheet basic 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
              <link rel=\\"stylesheet\\" href=\\"1.css\\" integrity=\\"sha384-SI8Lww7SI7ojkIK+OHWMVYWT/SI2f5rkGslJMkk/3C9mILFt0yMwTK3mlZL+BTmO\\" crossorigin=\\"anonymous\\">
          
        </body></html>",
}
`;

exports[`stylesheet crossorigin attribute 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"1.css\\" integrity=\\"sha384-SI8Lww7SI7ojkIK+OHWMVYWT/SI2f5rkGslJMkk/3C9mILFt0yMwTK3mlZL+BTmO\\" crossorigin=\\"use-credentials\\">
          
        </body></html>",
}
`;

exports[`stylesheet invalid hash algorithm 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"1.css\\" integrity=\\"sha1-dph/sbf7TTz0ynQdi1KxFDfv1Sg=\\" crossorigin=\\"anonymous\\">
          
        </body></html>",
}
`;

exports[`stylesheet invalid link 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"invalid.css\\">
          
        </body></html>",
}
`;

exports[`stylesheet multiple hash algorithms 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"1.css\\" integrity=\\"sha256-CWcnFgaWAReWDakUpF4oUc3hOL5tBSLTPTAxbLA5sbg= sha384-SI8Lww7SI7ojkIK+OHWMVYWT/SI2f5rkGslJMkk/3C9mILFt0yMwTK3mlZL+BTmO sha512-OSSE/t48EtyjumYELSZNOWlSBXEq7qRMfHWCCtb0bbq4fDCWr07dsRtMVfk3Bi6Hi4H/oK1ba703f5bvlCIIvA== whirlpool-ysTPMKhdX5R3Rf4gggA3rE7bwgcs0vUkvWw5w4b5FOJtcxcgP6J0n2veREIJjBb9ptUrsSVi54ugn2JdLGZekA==\\" crossorigin=\\"anonymous\\">
          
        </body></html>",
}
`;

exports[`stylesheet remote 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"https://raw.githubusercontent.com/JonasKruckenberg/rollup-plugin-sri/master/test/fixtures/1.css\\">
          
        </body></html>",
}
`;

exports[`stylesheet uppercase hash algorithms 1`] = `
Object {
  "fileName": "index.html",
  "source": "<!DOCTYPE html><html lang=\\"en\\"><head></head><body>
            <link rel=\\"stylesheet\\" href=\\"1.css\\" integrity=\\"sha256-CWcnFgaWAReWDakUpF4oUc3hOL5tBSLTPTAxbLA5sbg= sha384-SI8Lww7SI7ojkIK+OHWMVYWT/SI2f5rkGslJMkk/3C9mILFt0yMwTK3mlZL+BTmO sha512-OSSE/t48EtyjumYELSZNOWlSBXEq7qRMfHWCCtb0bbq4fDCWr07dsRtMVfk3Bi6Hi4H/oK1ba703f5bvlCIIvA==\\" crossorigin=\\"anonymous\\">
          
        </body></html>",
}
`;
